server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala

```diff
@@ -1,36 +1,42 @@
 package cmwell.tools.data.sparql
 
-import cmwell.tools.data.downloader.consumer.Downloader.Token
-
+import cmwell.tools.data.utils.akka.stats.DownloaderStats.DownloadStats
+import cmwell.util.http.SimpleResponse.Implicits.UTF8StringHandler
+import io.circe._, io.circe.parser._
 import scala.concurrent.ExecutionContext
 
 object StpUtil {
   def headerString(header: (String, String)): String = header._1 + ":" + header._2
 
   def headersString(headers: Seq[(String, String)]): String = headers.map(headerString).mkString("[", ",", "]")
 
-
   def extractLastPart(path: String) = {
     val p = if (path.endsWith("/")) path.init
     else path
     val (_, name) = p.splitAt(p.lastIndexOf("/"))
-    name.tail.init
+    name.tail
   }
 
   def readPreviousTokens(baseUrl: String, path: String, format: String)(implicit context : ExecutionContext) = {
-    import cmwell.util.http.SimpleResponse.Implicits.UTF8StringHandler
-
-    cmwell.util.http.SimpleHttpClient.get(s"http://$baseUrl$path/tokens?op=stream&recursive&format=$format")
-      .map(
-        _.payload.split("\n")
-          .map(_.split(" "))
-          .collect { case Array(s, p, o, _) =>
-            val token = if (o.startsWith("\"")) o.init.tail else o
-            extractLastPart(s) -> token
+    cmwell.util.http.SimpleHttpClient.get(s"http://$baseUrl$path/tokens?op=stream&recursive&format=json")
+      .map(response =>{
+        response.payload.lines.map( {row =>
+          parse(row) match {
+            case Left(parseFailure@ParsingFailure(_, _)) => throw parseFailure
+            case Right(json) => {
+
+              val token = json.hcursor.downField("fields").downField("token").values.get(0).asString.get
+
+              val receivedInfotons : Option[DownloadStats] = json.hcursor.downField("fields").downField("receivedInfotons").downArray.as[Long].toOption.map{ value =>
+                DownloadStats(receivedInfotons=value)
+              }
+
+              val sensor = extractLastPart(json.hcursor.downField("system").get[String]("path").toOption.get)
+              sensor -> (token,receivedInfotons)
+            }
           }
-          .foldLeft(Map.empty[String, Token])(_ + _)
-      )
+        }).foldLeft(Map.empty[String,TokenAndStatistics])(_ + _)
```

Comment:
use:
`.foldLeft(Map.newBuilder[String,TokenAndStatistics])(_.+=(_)).result()`
instead so you won't build immutable Map in memory for each iteration....

Reply:
Addressed this...